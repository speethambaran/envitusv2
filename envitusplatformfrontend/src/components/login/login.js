import React from 'react';
import { Form, Button, Modal, Container, Col, Row } from 'react-bootstrap';
import { authUser, getUserRole, regUser } from '../../services/userApi';
import { Redirect } from "react-router-dom";
import { connect } from 'react-redux';

import { superAdmin, admin, supervisor, userFormData, userFrmValSts } from '../../constants';
import { userPrivilegeRole } from '../../action/userAction';
import { intialaizeForm } from '../../action/formAction';

import { toastr } from 'react-redux-toastr';
import Cards from '../card/cards';
import RegisterForm from './registerForm.js';
import './login.scss';

export class Login extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            userName: React.createRef(),
            password: React.createRef(),
            userRole: '',
            showModal: false,
        };

        this.props.intialaizeForm({
            data: userFormData,
            id: '',
            deviceFrmValSts: { ...userFrmValSts },
            updateType: 'add'
        });

        this.showModal = this.showModal.bind(this);
        this.handleClose = this.handleClose.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
        this.register = this.register.bind(this);
    }

    async handleSubmit(event) {
        event.preventDefault();
        const logginDetails = {
            userName: this.state.userName.current.value,
            password: this.state.password.current.value
        }
        try {
            const status = await this.props.authUser(logginDetails);
            const uname = logginDetails.userName;
            const roleArray = await this.props.getUserRole(uname);

            if (roleArray) {
                if (JSON.stringify(roleArray) === JSON.stringify(superAdmin)) {
                    this.state.userRole = 'superAdmin';
                }
                else if (JSON.stringify(roleArray) === JSON.stringify(admin)) {
                    this.state.userRole = 'Administrator';
                }
                else if (JSON.stringify(roleArray) === JSON.stringify(supervisor)) {
                    this.state.userRole = 'Supervisor';
                }
                else {
                    this.state.userRole = 'Operator';
                }
            }
            if (this.state.userRole !== '') {
                sessionStorage.setItem("userRole", this.state.userRole);
                this.props.userPrivilegeRole(this.state.userRole);
            }

            if (!status) {
                toastr.error('Oops !!', 'Invalid Username or Password');
            }
        } catch (err) {
            toastr.error('Oops !!', 'Some error Occured');
        }
    }

    showModal() {
        this.setState({ showModal: true });
    }

    handleClose() {
        this.props.intialaizeForm({
            data: userFormData,
            id: '',
            deviceFrmValSts: { ...userFrmValSts },
            updateType: 'add'
        });
        this.setState({ showModal: false });
    }

    async register(e) {
        e.preventDefault();
        try {
            this.props.formData.data.activated = true;
            this.props.formData.data.creationLog = {
                user: this.props.formData.data.userName,
                date: new Date().toISOString().replace(/T/, ' ').replace(/\..+/, '')
            };
            this.props.formData.data.devices = [];
            this.props.formData.data.role = "Operator";
            const status = await this.props.regUser(this.props.formData.data);
            if (!status) {
                toastr.error('Oops !!', 'Some Error Occured');
            } else {
                toastr.success('Success !!', 'User Registered. Now Login');
            }
        } catch (err) {
            toastr.error('Oops !!', 'Some Error Occured');
        }
        this.handleClose();
    }

    render() {
        if (this.props.isLoggedIn) {
            return <Redirect data-test="redirectTag" to='/dashboard' />;
        }

        const cardContent = {
            head: <h1>Login</h1>, divItem:
                <div>
                    <Form data-test="loginForm" onSubmit={this.handleSubmit}>
                        <Form.Group controlId="formBasicEmail">
                            <Form.Control data-test="loginFormUname" type="text" ref={this.state.userName}
                                placeholder="Username" required
                            />
                        </Form.Group>
                        <Form.Group controlId="formBasicPassword">
                            <Form.Control data-test="loginFormPswd" type="password" ref={this.state.password}
                                placeholder="Password" required
                            />
                        </Form.Group>
                        <Button variant="primary" type="submit" className="w-100">
                            Login
                    </Button>
                    </Form>
                    <div className="mt-5">
                        <small>Not a User ?</small>
                        <summary>
                            <strong className="ml-2 hover-item" onClick={this.showModal}>
                                Register
                        </strong>
                        </summary>
                    </div>
                </div>
        };

        return (
            <div>
                <Cards content={cardContent} styles="border-0 shadow" />
                <Modal aria-labelledby="contained-modal-title-vcenter" size="md" scrollable="true"
                    show={this.state.showModal} onHide={this.handleClose}
                >
                    <Modal.Header closeButton>
                        <Modal.Title id="contained-modal-title-vcenter">
                            Register User
                        </Modal.Title>
                    </Modal.Header>
                    <Modal.Body>
                        <Container>
                            <Form id="myform" onSubmit={this.register}>
                                <RegisterForm />
                            </Form>
                            <Row>
                                <Col className="pr-1">
                                    <Button className="w-100" variant="danger" onClick={this.handleClose}>Close</Button>
                                </Col>
                                <Col className="pl-1">
                                    <Button className="w-100" form="myform" variant="primary" type="submit">Save</Button>
                                </Col>
                            </Row>
                        </Container>
                    </Modal.Body>
                </Modal>
            </div>
        );
    }
}


export function mapStateToProps(state) {
    return ({
        isLoggedIn: state.user.isLoggedIn,
        formData: state.formData,
    });
}

export function mapDispatchToProps(dispatch) {
    return {
        authUser: (logginDetails) => {
            return dispatch(authUser(dispatch, logginDetails))
        },
        getUserRole: (uname) => {
            return dispatch(getUserRole(dispatch, uname))
        },
        userPrivilegeRole: (role) => {
            return dispatch(userPrivilegeRole(role))
        },
        regUser: (data) => {
            return dispatch(regUser(dispatch, data))
        },
        intialaizeForm: (data) => {
            return dispatch(intialaizeForm(data))
        }
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(Login);
