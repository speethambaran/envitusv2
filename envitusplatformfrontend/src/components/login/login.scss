@import '../../styles/index.scss';

.hover-item:hover {
    color: $base-hover-color;
}