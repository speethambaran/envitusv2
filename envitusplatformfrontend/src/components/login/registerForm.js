import React from 'react';
import { Form } from 'react-bootstrap';
import { userFrmValRgex } from '../../constants';
import { connect } from 'react-redux';
import { updateFormValue, updateFormValidation } from '../../action/formAction';

class RegisterForm extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            
        }
        
        this.onChange = this.onChange.bind(this);
    }

    onChange(e) {
        this.props.updateFormValue({[e.target.name]: e.target.value});
        if(userFrmValRgex[e.target.name] && !userFrmValRgex[e.target.name].test(e.target.value)) {
            this.props.updateFormValidation({[e.target.name]: true});
        } else {
            this.props.updateFormValidation({[e.target.name]: false});
        }
    }

    render() {
        return (
            <React.Fragment>
                <Form.Group controlId="name">
                    <Form.Control value={this.props.userFrmDta.name}
                        isInvalid={this.props.userFrmValSts.name} placeholder="Name"
                        onChange={this.onChange} name="name" type="text" required
                    />
                    <Form.Control.Feedback type="invalid">Invalid Name</Form.Control.Feedback>
                </Form.Group>
                <Form.Group controlId="email">
                    <Form.Control value={this.props.userFrmDta.email}
                        isInvalid={this.props.userFrmValSts.email} placeholder="Email"
                        onChange={this.onChange} name="email" type="email" required
                    />
                    <Form.Control.Feedback type="invalid">Invalid Email</Form.Control.Feedback>
                </Form.Group>
                <Form.Group controlId="contact">
                    <Form.Control value={this.props.userFrmDta.contact}
                        isInvalid={this.props.userFrmValSts.contact} placeholder="Contact"
                        onChange={this.onChange} name="contact" type="text" required
                    />
                    <Form.Control.Feedback type="invalid">Invalid Contact</Form.Control.Feedback>
                </Form.Group>
                <Form.Group controlId="userName">
                    <Form.Control value={this.props.userFrmDta.userName} placeholder="Username"
                        isInvalid={this.props.userFrmValSts.userName}
                        onChange={this.onChange} name="userName" type="text" required
                    />
                    <Form.Control.Feedback type="invalid">Invalid Username</Form.Control.Feedback>
                </Form.Group>
                <Form.Group controlId="password">
                    <Form.Control value={this.props.userFrmDta.password}
                        isInvalid={this.props.userFrmValSts.password} placeholder="Password"
                        onChange={this.onChange} name="password" type="password" required
                    />
                    <Form.Control.Feedback type="invalid">
                        Invalid Password !! Should be atleast 8 in length and should contain atleast 1 alphabet and a number
                    </Form.Control.Feedback>
                </Form.Group>
            </React.Fragment>
        );
    }
}



function mapStateToProps(state) {
    return ({
        userFrmDta: state.formData.data,
        userFrmValSts: state.formData.deviceFrmValSts
    });
}

function mapDispatchToProps(dispatch) {
    return ({
        updateFormValue: (data) => {
            return dispatch(updateFormValue(data))
        },
        updateFormValidation: (data) => {
            return dispatch(updateFormValidation(data))
        }     
    });
}

export default connect(mapStateToProps, mapDispatchToProps)(RegisterForm);
