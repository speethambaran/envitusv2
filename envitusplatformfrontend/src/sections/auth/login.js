import React from 'react';
import { useForm } from "react-hook-form";
import { authUser } from '../../services/v1/authApi';
import { useDispatch, useSelector } from 'react-redux';
import { Redirect } from "react-router-dom";

const Login = () => {
	const { register, handleSubmit } = useForm();
	const dispatch = useDispatch();
	const { isLoggedIn } = useSelector(
		state => ({
			isLoggedIn: state.user.isLoggedIn
		})
	);

	const onSubmitForm = (formData) => {
		dispatch(authUser(formData));
	}

	if (isLoggedIn) {
		return <Redirect to='/dashboard' />;
	} else {
		return (
			<div className="content content-full">
				<div className="px-30 py-10 logo">
					<img src={process.env.PUBLIC_URL + process.env.REACT_APP_HOME_LOGO} alt="logo" />
					<h1 className="h4 font-w700 mt-30 mb-10">Welcome to Your Dashboard</h1>
				</div>

				<form className="js-validation-signin px-30" onSubmit={handleSubmit(onSubmitForm)}>
					<div className="form-group row">
						<div className="col-12">
							<div className="form-material floating">
								<input type="text" className="form-control" id="login-username" name="username" ref={register({ required: true })} />
								<label htmlFor="login-username">Username</label>
							</div>
						</div>
					</div>
					<div className="form-group row">
						<div className="col-12">
							<div className="form-material floating">
								<input type="password" className="form-control" id="login-password"
									name="password" ref={register({ required: true })} />
								<label htmlFor="login-password">Password</label>
							</div>
						</div>
					</div>
					<div className="form-group">
						<button className="btn btn-sm btn-hero btn-alt-primary">
							<i className="si si-login mr-10"></i> Sign In
                                        </button>
					</div>
				</form>
			</div>
		)
	}
}

export default Login;
