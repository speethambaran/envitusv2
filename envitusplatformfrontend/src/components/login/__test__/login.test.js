import React from 'react';
import { shallow } from 'enzyme';
import { Login, mapStateToProps, mapDispatchToProps } from './../login';
import { findByDataTest } from './../../../Utils';

const preConfig = (props={}) => {
    const component = shallow(<Login {...props} />);
    return component;
};

describe('Login Test', () => {
    
    describe('Unauthenticated User Stays in Home', () => {
        let component;
        beforeEach(() =>{
            const props = {
                isLoggedIn: false
            };
            component = preConfig(props); 
        });

        it('render login form without errors', () => {
            const wrapper = findByDataTest(component, 'loginForm');
            expect(wrapper.length).toBe(1);
        });

        it('do not render redirect tag', () => {
            const wrapper = findByDataTest(component, 'redirectTag');
            expect(wrapper.length).toBe(0);
        });
    });

    describe('Authenticated User redirected to Dashboard', () => {
        let component;
        beforeEach(() =>{
            const props = {
                isLoggedIn: true
            };
            component = preConfig(props); 
        });

        it('render redirect tag without errors', () => {
            const wrapper = findByDataTest(component, 'redirectTag');
            expect(wrapper.length).toBe(1);
        });
    });

    describe('Rendering Input types correctly', () => {
        let component;
        beforeEach(() => {
            component = preConfig();
        });

        it('Username Render', () => {
            const wrapper = findByDataTest(component , 'loginFormUname');
            expect(wrapper.prop('type')).toBe('text');
        });

        it('Password Render', () => {
            const wrapper = findByDataTest(component , 'loginFormPswd');
            expect(wrapper.prop('type')).toBe('password');
        });
    });

    describe('Authentication', () => {
        it('snapshot', () => {
            const wrapper = shallow(<Login/>);   
            expect(wrapper).toMatchSnapshot();
        });

        it('Login with correct credential', () => {
            const loginComponent = shallow(<Login />);
            expect(loginComponent.find('Form').length).toBe(1);
            loginComponent.find('Form').simulate('submit', loginComponent.setState({userName: 'sudo', password: 'sudo123', showWarning : true }));
            expect(loginComponent.state("showWarning")).toBe(true);
        });

        it('cannot Login with correct credential', () => {
            const loginComponent = shallow(<Login />);
            expect(loginComponent.find('Form').length).toBe(1);
            loginComponent.find('Form').simulate('submit', loginComponent.setState({userName: 'sudo', password: 'sudo1234', showWarning : false }));
            expect(loginComponent.state("showWarning")).toBe(false);
        });
    });

    describe('State Dispach Prop', () => {
        it('should show previousLog in State', () => {
            const initialState = {
                user: {
                    isLoggedIn: true
                }
            };
            expect(mapStateToProps(initialState).isLoggedIn).toBe(true);
        });

        it('authUser returns function', () => {
            const dispatch = jest.fn();
            mapDispatchToProps(dispatch).authUser();
            expect(typeof dispatch.mock.calls[0][0]).toBe('function');
        });

        it('getUserRole returns function', () => {
            const dispatch = jest.fn();
            mapDispatchToProps(dispatch).getUserRole();
            expect(typeof dispatch.mock.calls[0][0]).toBe('function');
        });

        it('userPrivilegeRole returns function', () => {
            const dispatch = jest.fn();
            mapDispatchToProps(dispatch).userPrivilegeRole();
            expect(typeof dispatch.mock.calls[0][0]).toBe('object');
        });

    });

    describe('function test', () => {
        it("handleSubmit authUser return false: promise fulfilled : superAdmin", async () => {
            let component;
            const mockAuthUser = jest.fn();
            mockAuthUser.mockReturnValueOnce(false); 
            
            const mockGetUserRole = jest.fn();
            mockGetUserRole.mockReturnValueOnce([true, true, true, true, true]);

            const mockUserPrivilegeRole = jest.fn();
            
            const mockEvent = {
                preventDefault: jest.fn()
            };

            const props = {
                authUser: mockAuthUser,
                getUserRole: mockGetUserRole,
                userPrivilegeRole: mockUserPrivilegeRole
            };

            component = preConfig(props);

            component.setState({
                userName: {current: {value: 'uname'}},
                password: {current: {value: 'pwd'}}              
            });

            return(component.instance().handleSubmit(mockEvent)).then(() => {
                expect(mockAuthUser.mock.calls.length).toBe(1);
                expect(mockGetUserRole.mock.calls.length).toBe(1);
                expect(component.state("userRole")).toBe('superAdmin');
                expect(mockUserPrivilegeRole.mock.calls.length).toBe(1);
                expect(component.state("showWarning")).toBe(true);
                expect(component.state("warningType")).toBe('warning');
            });
        });

        it("handleSubmit authUser return false: promise fulfilled : Administrator", async () => {
            let component;
            const mockAuthUser = jest.fn();
            mockAuthUser.mockReturnValueOnce(false); 
            
            const mockGetUserRole = jest.fn();
            mockGetUserRole.mockReturnValueOnce([true, true, true, false, true]);

            const mockEvent = {
                preventDefault: jest.fn()
            };

            const props = {
                authUser: mockAuthUser,
                getUserRole: mockGetUserRole
            };

            component = preConfig(props);

            component.setState({
                userName: {current: {value: 'uname'}},
                password: {current: {value: 'pwd'}}              
            });

            return(component.instance().handleSubmit(mockEvent)).then(() => {
                expect(mockAuthUser.mock.calls.length).toBe(1);
                expect(mockGetUserRole.mock.calls.length).toBe(1);
                expect(component.state("userRole")).toBe('Administrator');
            });
        });

        it("handleSubmit authUser return false: promise fulfilled : Supervisor", async () => {
            let component;
            const mockAuthUser = jest.fn();
            mockAuthUser.mockReturnValueOnce(false); 
            
            const mockGetUserRole = jest.fn();
            mockGetUserRole.mockReturnValueOnce([false, true, true, false, false]);

            const mockEvent = {
                preventDefault: jest.fn()
            };

            const props = {
                authUser: mockAuthUser,
                getUserRole: mockGetUserRole
            };

            component = preConfig(props);

            component.setState({
                userName: {current: {value: 'uname'}},
                password: {current: {value: 'pwd'}}              
            });

            return(component.instance().handleSubmit(mockEvent)).then(() => {
                expect(mockAuthUser.mock.calls.length).toBe(1);
                expect(mockGetUserRole.mock.calls.length).toBe(1);
                expect(component.state("userRole")).toBe('Supervisor');
            });
        });

        it("handleSubmit authUser return false: promise fulfilled : Operator", async () => {
            let component;
            const mockAuthUser = jest.fn();
            mockAuthUser.mockReturnValueOnce(false); 
            
            const mockGetUserRole = jest.fn();
            mockGetUserRole.mockReturnValueOnce([false, true, false, false, false]);

            const mockEvent = {
                preventDefault: jest.fn()
            };

            const props = {
                authUser: mockAuthUser,
                getUserRole: mockGetUserRole
            };

            component = preConfig(props);

            component.setState({
                userName: {current: {value: 'uname'}},
                password: {current: {value: 'pwd'}}              
            });

            return(component.instance().handleSubmit(mockEvent)).then(() => {
                expect(mockAuthUser.mock.calls.length).toBe(1);
                expect(mockGetUserRole.mock.calls.length).toBe(1);
                expect(component.state("userRole")).toBe('Operator');
            });
        });

        it("handleSubmit authUser return true: promise fulfilled + no roleArray", async () => {
            let component;
            const mockAuthUser = jest.fn();
            mockAuthUser.mockReturnValueOnce(true); 

            const mockGetUserRole = jest.fn();
            mockGetUserRole.mockReturnValueOnce(false);

            const mockUserPrivilegeRole = jest.fn();

            const mockEvent = {
                preventDefault: jest.fn()
            };

            const props = {
                authUser: mockAuthUser,
                getUserRole: mockGetUserRole,
                userPrivilegeRole: mockUserPrivilegeRole
            };

            component = preConfig(props);

            component.setState({
                userName: {current: {value: 'uname'}},
                password: {current: {value: 'pwd'}}              
            });

            return(component.instance().handleSubmit(mockEvent)).then(() => {
                expect(mockAuthUser.mock.calls.length).toBe(1);                
                expect(mockGetUserRole.mock.calls.length).toBe(1);
                expect(component.state("userRole")).toBe('');
                expect(mockUserPrivilegeRole.mock.calls.length).toBe(0);
                expect(component.state("showWarning")).toBe(false);
            });
        });

        it("handleSubmit authUser return false: promise rejected", async () => {
            let component;
            const mockAuthUser = jest.fn();
            mockAuthUser.mockReturnValueOnce(false);  
            
            const mockEvent = {
                preventDefault: jest.fn()
            };

            const props = {
                authUser: mockAuthUser
            };

            component = preConfig(props);

            component.setState({
                userName: {current: {value: 'uname'}},
                password: {current: {value: 'pwd'}}              
            });

            return(component.instance().handleSubmit(mockEvent)).catch(e => {
                expect(e).toMatch('error');
                expect(mockAuthUser.mock.calls.length).toBe(0);
                expect(component.state("showWarning")).toBe(true);
                expect(component.state("warningType")).toBe('danger');
                expect(component.state("warningMsg")).toBe(null);
            })
        });
    });
});