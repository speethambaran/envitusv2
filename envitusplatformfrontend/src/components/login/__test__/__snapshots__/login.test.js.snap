// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Test Authentication snapshot 1`] = `
<Form
  data-test="loginForm"
  inline={false}
  onSubmit={[Function]}
>
  <FormGroup
    controlId="formBasicEmail"
  >
    <FormLabel
      column={false}
      srOnly={false}
    >
      Username
    </FormLabel>
    <FormControl
      data-test="loginFormUname"
      placeholder="Enter email"
      required={true}
      type="text"
    />
  </FormGroup>
  <FormGroup
    controlId="formBasicPassword"
  >
    <FormLabel
      column={false}
      srOnly={false}
    >
      Password
    </FormLabel>
    <FormControl
      data-test="loginFormPswd"
      placeholder="Password"
      required={true}
      type="password"
    />
  </FormGroup>
  <AlertBox
    show={false}
    text=""
    variant=""
  />
  <Button
    active={false}
    disabled={false}
    type="submit"
    variant="primary"
  >
    Login
  </Button>
</Form>
`;
